File: lru_cache.cpp
// https://leetcode.com/problems/lru-cache/
#include <unordered_map>
#include <list>
#include <utility>

class LRUCache {
private:
    typedef std::pair<int,int> cachePairType;
    typedef std::unordered_map <int, std::list<cachePairType>::iterator> cacheMapType;
    
    std::list <cachePairType> cacheList;
    cacheMapType cacheMap;
    int cacheSize;
    
public:
    
    LRUCache(int capacity) {
        cacheSize = capacity;
    }
    
    int get(int key) {
        cacheMapType::iterator cacheIt = cacheMap.find(key);
        
        if (cacheIt != cacheMap.end())
        {
            cachePairType item = *cacheIt->second;
            
            // Update the list
            cacheList.erase(cacheIt->second);
            cacheList.push_front(item);
            
            // Update the map with the new iterator
            cacheIt->second = cacheList.begin();
            
            return item.second;
        }
        else
        {
            return -1;
        }
        
    }
    
    void put(int key, int value) {
        std::pair item(key,value);
        cacheMapType::iterator cacheIt = cacheMap.find(key);
        
        if (cacheIt != cacheMap.end())
        {
            // Update the list
            cacheList.erase(cacheIt->second);
            cacheList.push_front(item);
            
            // Update the map with the new iterator
            cacheIt->second = cacheList.begin();
        }
        else
        {
            if(cacheList.size() >= cacheSize)
            {
                // Remove the item from both the map and list
                cachePairType evictedItem = cacheList.back();
                cacheMap.erase(cacheMap.find(evictedItem.first));
                cacheList.pop_back();
                
            }
            
            // Insert item in list and update map
            cacheList.push_front(item);
            cacheMap[key] = cacheList.begin();
        }
        
        
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

File: lru_cache.py
class Node:
    def __init__(self,data):
        self.data = data
        self.prev = None
        self.next = None
        
class DoublyLinkedList:
    def __init__(self):
        self.head = Node(None)
        self.tail = Node(None)
        
        self.head.next = self.tail
        self.tail.prev = self.head
        
    def append(self, node):
        node.next = self.tail
        node.prev = self.tail.prev
        
        self.tail.prev.next = node # set the old tail node to point to the new tail
        self.tail.prev = node # set pseudo tail to point to new tail
        
        return node
    
    def removeNode(self,node):
        tempNext = node.next
        tempPrev = node.prev
        
        tempPrev.next = tempNext
        tempNext.prev = tempPrev
            
        return node
    
    def pop(self):
        node = self.head.next
        self.removeNode(node)
        return node
    
class cacheObj:
    def __init__(self,value,key):
        self.val = value
        self.key = key
        
class LRUCache:

    def __init__(self, capacity: int):
        self.currSize = 0
        self.maxSize = capacity
        self.cacheMap = {} # store index of keys in list
        self.cacheList = DoublyLinkedList()
        

    def get(self, key: int) -> int:
        if key in self.cacheMap:
            # retrieve the value
            itemNode = self.cacheMap[key]
            self.cacheList.removeNode(itemNode)
            
            # update internal data structures
            self.cacheList.append(itemNode)
            self.cacheMap[key] = itemNode
            
            return itemNode.data.val
        else:
            return -1
            
    def put(self, key: int, value: int) -> None:
        if key in self.cacheMap:
            node = self.cacheMap[key]
            node.data.val = value
            self.cacheList.removeNode(node)
            self.cacheList.append(node)

        else:
            if self.currSize >= self.maxSize:
                evictedItem = self.cacheList.pop().data
                self.currSize -= 1
                del self.cacheMap[evictedItem.key]
            
            
            node = self.cacheList.append(Node(cacheObj(value,key)))
            self.cacheMap[key] = node
            self.currSize += 1
        
# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
